Add information about PodReplacementPolicy in Job API (#41745)
kubernetes/website: content/en/docs/concepts/workloads/controllers/job.md

ensures that deleted pods have their finalizers removed by the Job controller.
{{< /note >}}

{{< note >>}}

> Since Kubernetes v1.28, when pod failure policy is used, the Job controller will recreate

Reviewer: ```suggestion
Since Kubernetes v1.28, when pod failure policy is used, the Job controller recreates
```

terminating Pods only once these Pods reach the terminal `Failed` phase. This behavior is similar

> to `podRecreationPolicy: Failed`. For more information, see [Pod replacement policy](#pod-replacement-policy).

Reviewer 2: ```suggestion
to `podReplacementPolicy: Failed`. For more information, see [Pod replacement policy](#pod-replacement-policy).
```

there are also other instances of `podRecreationPolicy` that should be updated: https://github.com/search?q=repo%3Akubernetes%2Fkubernetes%20podRecreationPolicy&type=code

Author: Wow! Thanks for finding that.  I'll open up a PR today.

I'll also update the code.

> Since Kubernetes v1.28, when pod failure policy is used, the Job controller recreates
> terminating pods only once they reach the terminal `Failed` phase. This behavior is analogous
> to when using `podRecreationPolicy: Failed`, see [pod replacement policy](#pod-replacement-policy) for more details.
> {{< /note >}}

Reviewer 3: ```suggestion
Starting with Kubernetes v1.28, when Pod failure policy is used, the Job controller recreates
terminating Pods only once these Pods reach the terminal `Failed` phase. This behavior is similar
to `podRecreationPolicy: Failed`. For more information, see [Pod replacement policy](#pod-replacement-policy).
{{< /note >}}
```

Author: updated.

Use cases for elastic Indexed Jobs include batch workloads which require 
scaling an indexed Job, such as MPI, Horovord, Ray, and PyTorch training jobs.

### Recreation only when Pod has been fully terminated

{{< feature-state for_k8s_version="v1.28" state="alpha" >}}

> For distributed jobs, it is common to have many replicas running in a job.  

Reviewer: I would suggest to rewrite this paragraph a little keeping in mind the guidance style suggestions: https://kubernetes.io/docs/contribute/style/style-guide/. In particular "use present tense" or "Address the reader as "you"". Maybe something along the lines (but feel free to rephrase also this, just consider as a general direction):
```
By default, the Job controller recreates pods as soon they are either failed or have a deletion timestamp. 

This means, that you may have two running at the same time, both corresponding to the same pod. In that case, one is terminating, while the other is starting up. 

This behavior might be problematic in some situations, including:
- when using a framework which requires only a single pod per index (for example TF, JAX) <- maybe with link to the issue
- when using a cluster autoscaler, which could increase the size of the cluster for only a short while during pod termination.

If this behavior is problematic for your Jobs, you may choose to create replacement pods only when the terminating pod is fully terminal. To do this, set the `.spec.podReplacementPolicy: Failed`.

Some frameworks will require uniqueness for the pod index.  
In cases like this, it is beneficial to only recreate pods
based once the pod has been terminated.

> When the `JobPodReplacementPolicy`[feature gate](/docs/reference/command-line-tools-reference/feature-gates/) is set, you will be able to specify `PodReplacementPolicy` to `Failed`.  

Reviewer 4: The PR title says PodRecreationPolicy. Should that be updated?

Author: PR title was wrong.  Good catch on that one.  I updated title to say `PodReplacementPolicy`.

This will only recreate pods once they are terminated.  
Default behavior will be to recreate upon deletion (`DeletionTimestamp != nil`).

```yaml
kind: Job
metadata:
  name: new
  ...
spec:

>   PodReplacementPolicy: Failed

Reviewer 5: Did you mean:
```suggestion
  podReplacementPolicy: Failed
```

?

Author: Yep!

  ...
```

> To improve debuggability/observability for pods that are being deleted (`DeletionTimestamp != nil`) we include a new field in the JobStatus called `terminating`.  

Reviewer: Again, I would like to use more of active voice. Maybe, "When using the podReplacementPolicy: Failed, you can inspect the number of terminating pods in the status..."

Author: Updated.

This will report the number pods that are currently terminating and is easily viewable in the status.

```shell
kubectl get jobs/myjob -o yaml
```

```yaml
apiVersion: batch/v1
kind: Job
# .metadata and .spec omitted
status:
  terminating: 1 # if pod is terminating
```

> In cases where `PodFailurePolicy` is enabled, we will assume `Failed` as the default and only allowed value.  

Reviewer: I think it would also be beneficial to x-reference this section (with links) from the pod failure policy, and vice versa.

> In cases where `PodFailurePolicy` is enabled, we will assume `Failed` as the default and only allowed value.  

Reviewer: Aside from the style guide, this `will assume` makes me wonder if the docs are up to date? Do we currently go with the approach of implicitly assuming `Failed`, or we actually materialize the value?

Author: I corrected this.  We set default values based on `JobPodRecreationPolicy` gate.  I think in cases of PodFailurePolicy we default to Failed.

- when using a cluster autoscaler, which could increase the size of the cluster for only a short while during pod termination.

If this behavior is problematic for your Jobs, you may choose to create replacement pods only when the terminating pod is fully terminal. To do this, set the `.spec.podReplacementPolicy: Failed`.
When the `JobPodReplacementPolicy`[feature gate](/docs/reference/command-line-tools-reference/feature-gates/) is set, you will be able to specify `PodReplacementPolicy` to `Failed`.  
This will only recreate pods once they are terminated.  
Default behavior will be to recreate upon deletion (`DeletionTimestamp != nil`).

```yaml
kind: Job
metadata:
  name: new
  ...
spec:
  podReplacementPolicy: Failed
  ...
```

When using the feature gate `JobPodReplacementPolicy`, you can inspect a new field in the JobStatus called `terminating`.  
This will report the number pods that are currently terminating and is easily viewable in the status.

```shell
kubectl get jobs/myjob -o yaml
```

```yaml
apiVersion: batch/v1
kind: Job
# .metadata and .spec omitted
status:
  terminating: 1 # if pod is terminating
```

> In cases where `JobPodReplacementPolicy` and [PodFailurePolicy]((/docs/tasks/job/pod-failure-policy/)) is enabled, a Job will have a default (and only this value is allowed) value of `Failed`.

Reviewer: ```suggestion
In cases where `JobPodReplacementPolicy` and [PodFailurePolicy]((/docs/tasks/job/pod-failure-policy/)) are enabled, a Job will have a default (and only this value is allowed) value of `Failed`.
```

Author: Fixed.

> In cases where `JobPodReplacementPolicy` and [PodFailurePolicy]((/docs/tasks/job/pod-failure-policy/)) is enabled, a Job will have a default (and only this value is allowed) value of `Failed`.

Reviewer: Maybe we could also add a note to pod failure policy at the end, saying something like: 
```
Since Kubernetes v1.28, when pod failure policy is used, the Job controller will recreate
terminating pods only once they reach the terminal `Failed` phase. This behavior is analogous 
to when using `podRecreationPolicy: Failed`, see [pod recreation policy](#the link) for more details.
```

Author: Added.

> In cases where `JobPodReplacementPolicy` is disabled and `podFailurePolicy` is enabled, A Job will wait for terminating pods to be fully terminated before marking the pod as failed.  

Reviewer: ```suggestion
When `JobPodReplacementPolicy` is disabled and `podFailurePolicy` is enabled, then Job will wait for terminating pods to be fully terminated before marking the pod as failed.  
```

Use cases for elastic Indexed Jobs include batch workloads which require 
scaling an indexed Job, such as MPI, Horovord, Ray, and PyTorch training jobs.

> ### Pod Replacement Policy

Author: updated.

{{< feature-state for_k8s_version="v1.28" state="alpha" >}}

{{< note >}}

> You can only enable `PodReplacementPolicy` for Jobs if you set `JobPodReplacementPolicy`[feature gate](/docs/reference/command-line-tools-reference/feature-gates/) to true.

Reviewer 6: ```suggestion
You can only set `podReplacementPolicy` on Jobs if you enable the `JobPodReplacementPolicy` [feature gate](/docs/reference/command-line-tools-reference/feature-gates/).
```

Author: done.

{{< /note >}}

> By default, the Job controller recreates pods as soon they are either failed or terminating (have a deletion timestamp).

Reviewer 3: ```suggestion
By default, the Job controller recreates Pods as soon they either fail or are terminating (have a deletion timestamp).
```

Author: updated.

> This means that, at a given time, the number of running Pods for the Jobs can be greater than `parallelism` or, if using Indexed Jobs, more than one running Pod per index, if some of the Pods are terminating.

Reviewer 3: ```suggestion
This means that, at a given time, when some of the Pods are terminating, the number of running Pods for the Jobs can be greater than `parallelism` or greater than the running Pod per index (if using Indexed Jobs).
```

Not sure if my suggestion is still accurate, but the original sentence is hard to read

Author: I think you were very close.  I added one part that makes it correct.

> You may choose to create replacement pods only when the terminating pod is fully terminal (has `status.phase: Failed`). To do this, set the `.spec.podReplacementPolicy: Failed`.

Reviewer 3: ```suggestion
You may choose to create replacement pods only when the terminating pod is fully terminated (has `status.phase: Failed`). To do this, set the `.spec.podReplacementPolicy: Failed`.
```

> This will only recreate pods once they are terminated.  

Reviewer 3: I think you can omit this, but it's only a suggestion

Author: updated.

> Default behavior will be to recreate upon deletion (`DeletionTimestamp != nil`).

Reviewer 6: ```suggestion
The default policy is `FailedOrTerminating`, meaning that the control plane creates replacement Pods upon deletion (`DeletionTimestamp != nil`).
```

Author: Updated.

> The default policy is `FailedOrTerminating`, meaning that the control plane creates replacement Pods upon deletion (`DeletionTimestamp != nil`).

Reviewer 5: ```suggestion
The default policy is `FailedOrTerminating`, meaning that the control plane creates replacement Pods
immediately upon deletion (as soon as the control plane sees that a Pod for this Job has
`deletionTimestamp` set).
```

Author: Accepted and removed paragraph about `podFailurePolicy`

> The default policy is `FailedOrTerminating`, meaning that the control plane creates replacement Pods upon deletion (`DeletionTimestamp != nil`).

Reviewer 3: ```suggestion
The default policy is `FailedOrTerminating`, hence the control plane creates replacement Pods upon deletion (`DeletionTimestamp != nil`).
```

```yaml
kind: Job
metadata:
  name: new
  ...
spec:
  podReplacementPolicy: Failed
  ...
```

> You can inspect a new field in the JobStatus called `terminating`.  

Author: updated.

> This will report the number pods that are currently terminating and is easily viewable in the status.

Reviewer 3: ```suggestion
This shows the number pods that are currently terminating.
```
I'd remove the part " and is easily viewable in the status" because it's ambiguous

> This will report the number pods that are currently terminating and is easily viewable in the status.

Author: updated.

```shell
kubectl get jobs/myjob -o yaml
```

```yaml
apiVersion: batch/v1
kind: Job
# .metadata and .spec omitted
status:

>   terminating: 1 # if pod is terminating

Author: updated.

```

> When [PodFailurePolicy](#pod-failure-policy) is enabled, a Job will have a default (and only this value is allowed) value of `Failed`.

Author: Updated.

> When you use a [podFailurePolicy](#pod-failure-policy) in a Job, the Job will have a default `podReplacementPolicy` value of `Failed`, and this is the only policy allowed.
> If `JobPodReplacementPolicy` is disabled and `podFailurePolicy` is enabled, a Job will wait for terminating pods to be fully terminated before marking the pod as failed.

Reviewer 3: ```suggestion
When you use a [podFailurePolicy](#pod-failure-policy) in a Job, the Job has a default `podReplacementPolicy` value of `Failed`, and this is the only policy allowed.
If `JobPodReplacementPolicy` is disabled and `podFailurePolicy` is enabled, the Job waits for terminating pods to be fully terminated before marking the pod as failed.
```

Author: Yours and @sftim suggestions conflict here so maybe take a look over it and see if you are happy with the changes overall.

Reviewer 3: You can omit my suggestion. We recommend not using future tense, but it's okay if the context needs the future tense 😄

> If `JobPodReplacementPolicy` is disabled and `podFailurePolicy` is enabled, a Job will wait for terminating pods to be fully terminated before marking the pod as failed.

Author: updated.

Use cases for elastic Indexed Jobs include batch workloads which require 
scaling an indexed Job, such as MPI, Horovord, Ray, and PyTorch training jobs.

> ### Pod Replacement Policy {#pod-replacement-policy}

Author: Updated.

{{< feature-state for_k8s_version="v1.28" state="alpha" >}}

> By default, the Job controller recreates pods as soon they are either failed or have a deletion timestamp.

Author: Updated.

> This means, that you may have two running at the same time, both corresponding to the same pod. In that case, one is terminating, while the other is starting up.

Author: done.

> By default, the Job controller recreates Pods as soon they either fail or are terminating (have a deletion timestamp).

Reviewer 2: Does it make sense to make this more detailed here? That it recreates on success as well..

Reviewer 5: Why would Job recreate a Pod on success?

Reviewer 2: https://github.com/kubernetes/website/pull/41745#discussion_r1290132049

This means that, at a given time, when some of the Pods are terminating, the number of running Pods for the Jobs can be greater than `parallelism` or greater than one Pod per index (if using Indexed Jobs).

> You may choose to create replacement Pods only when the terminating Pod is fully terminal (has `status.phase: Failed`). To do this, set the `.spec.podReplacementPolicy: Failed`.

Reviewer 2: ```suggestion
You may choose to create replacement Pods only when the terminating Pod is fully terminal (`status.phase` has a value` Failed` or `Succeeded`). To do this, set the `.spec.podReplacementPolicy: Failed`.
```

Author: I don't think this is correct.  I don't know if you can actually have a terminating pod that is also succeeded.  I think the recreation would only happen if a pod is terminal with Failed.

Reviewer 5: For Pods that succeed, `.spec.podReplacementPolicy` makes no difference. We could fix this by adding a glossary entry for “Pod termination” and referring to that. Then, we can omit the text in parens here.

Reviewer: > I don't think this is correct. I don't know if you can actually have a terminating pod that is also succeeded. I think the recreation would only happen if a pod is terminal with Failed.

Correct, when `podReplacementPolicy: Failed` we only recreate pods when the previous is terminal AND failed. When `podReplacementPolicy: TerminatingOrFailed` (default) we also recreate pods when they are terminating, and they still may end up in `succeeded` state. This is confusing, but I think a little off here.

> For Pods that succeed, .spec.podReplacementPolicy makes no difference. 

Correct. 

> We could fix this by adding a glossary entry for “Pod termination” and referring to that. Then, we can avoid adding the text in parens here.

For now I would suggest to just rephrase it as "Pod is fully terminal in the `Failed` phase".

Reviewer 2: I had `completions > 1` in mind, which creates a new pod on success. I suppose it depends if we want to describe more generic behaviour of the controller, or if we want to make it specific to the feature.

These sentences evoke the generic behaviour:
- `By default, the Job controller recreates Pods as soon they either fail or are terminating (have a deletion timestamp).`
- `create replacement Pods only when`

> I don't think this is correct. I don't know if you can actually have a terminating pod that is also succeeded. I think the recreation would only happen if a pod is terminal with Failed.

yeah, I guess my suggestion makes it confusing, if by terminating pods we mean pods that they have received a termination signal.

>  When podReplacementPolicy: TerminatingOrFailed (default) we also recreate pods when they are terminating, and they still may end up in succeeded state. This is confusing, but I think a little off here.

we circumvent the problem here by defining it via deletionTimestamp

> For now I would suggest to just rephrase it as "Pod is fully terminal in the Failed phase".

+1, as it doesn't try to define terminal

Reviewer 2: okay

> If this behavior is problematic for your Jobs, you may choose to create replacement pods only when the terminating pod is fully terminal. To do this, set the `.spec.podReplacementPolicy: Failed`.

Author: done.

With no Pod failure policy defined for a Job, omitting the `podReplacementPolicy` field selects the

> `FailedOrTerminating` replacement policy:

Reviewer 2: ```suggestion
`TerminatingOrFailed`  replacement policy:
```

Author: Doh. I'll fix that also.

> When the `JobPodReplacementPolicy`[feature gate](/docs/reference/command-line-tools-reference/feature-gates/) is set, you will be able to specify `PodReplacementPolicy` to `Failed`.  

Author: Thanks!  This reads a lot better now.

Default behavior will be to recreate upon deletion (`DeletionTimestamp != nil`).

```yaml
kind: Job
metadata:
  name: new
  ...
spec:
  podReplacementPolicy: Failed
  ...
```

> When using the feature gate `JobPodReplacementPolicy`, you can inspect a new field in the JobStatus called `terminating`.  

Reviewer 6: No need to mention the feature gate again.

Author: done.

This will report the number pods that are currently terminating and is easily viewable in the status.

```shell
kubectl get jobs/myjob -o yaml
```

```yaml
apiVersion: batch/v1
kind: Job
# .metadata and .spec omitted
status:
  terminating: 1 # if pod is terminating
```

> When `JobPodReplacementPolicy` and [PodFailurePolicy](#pod-failure-policy) are enabled, a Job will have a default (and only this value is allowed) value of `Failed`.

Reviewer 6: Do not mention the JobPodReplacementPolicy feature gate again.

Author: I mention it in one place because i want to mention that this functionaly is on if PodFailurePolicy is enabled and feature is off.
